xls/dslx/bytecode/bytecode_interpreter_test.cc

```diff
@@ -1519,6 +1519,78 @@ fn main() {
   EXPECT_TRUE(value.IsUnit());
 }
 
+// GH 981
+// Assert_eq type deduction didn't fail in
+// the first iteration of the next function.
+// This test requires 2 iterations of next function,
+// which triggered assert_eq bug.
+TEST(BytecodeInterpreterTest, AssertEqFailProcs) {
+  constexpr std::string_view kProgram = R"(
+#[test_proc]
+proc BTester {
+    terminator: chan<bool> out;
+
+    init { (u32:0) }
+
+    config(terminator: chan<bool> out) {
+        (terminator,)
+    }
+
+    next(tok: token, state: u32) {
+        assert_eq(state, u32:0);
+        let tok = send_if(tok, terminator, state > u32:1, true);
+        (state + u32:1)
+    }
+})";
+
+  XLS_ASSERT_OK_AND_ASSIGN(auto temp_file,
+                           TempFile::CreateWithContent(kProgram, "_test.x"));
+  constexpr std::string_view kModuleName = "test";
+  ParseAndTestOptions options;
+  ::testing::internal::CaptureStderr();
+  absl::StatusOr<TestResult> result = ParseAndTest(
+      kProgram, kModuleName, std::string{temp_file.path()}, options);
+  std::string stdcerr(::testing::internal::GetCapturedStderr());
+  EXPECT_THAT(result, status_testing::IsOkAndHolds(TestResult::kSomeFailed));
+  EXPECT_THAT(stdcerr, HasSubstr("were not equal"));
+}
+
+// GH 981
+// Assert_lt type deduction didn't fail in
+// the first iteration of the next function.
+// This test requires 2 iterations of next function,
+// which triggered assert_lt bug.
+TEST(BytecodeInterpreterTest, AssertLtFailProcs) {
```

Comment:
I'm not sure if we need the extra test (as a regression in the codepath would likely trigger the other one as well)